### PR TITLE
vlan: add vlan.id keyword - v2

### PR DESCRIPTION
**Pull request by AkakiAlice**
Ticket: #1065

Make sure these boxes are checked accordingly before submitting your Pull Request -- thank you.

## Contribution style:
- [x] I have read the contributing guide lines at
   https://docs.suricata.io/en/latest/devguide/contributing/contribution-process.html

## Our Contribution agreements:
- [x] I have signed the Open Information Security Foundation contribution agreement at
   https://suricata.io/about/contribution-agreement/ (note: this is only required once)

## Changes (if applicable):
- [ ] I have updated the User Guide (in [doc/userguide/](https://github.com/OISF/suricata/tree/304271e63a9e388412f25f0f94a1a0da4bf619d9/doc/userguide)) to reflect the changes made
- [ ] I have updated the JSON schema (in [etc/schema.json](https://github.com/OISF/suricata/blob/304271e63a9e388412f25f0f94a1a0da4bf619d9/etc/schema.json)) to reflect all logging changes
      (including schema descriptions)
- [ ] I have created a ticket at
      https://redmine.openinfosecfoundation.org/projects/suricata/issues

Link to ticket: https://redmine.openinfosecfoundation.org/issues/1065

Describe changes:
- Introduce vlan.id keyword

SV_BRANCH= https://github.com/OISF/suricata-verify/pull/2134
Previous PR: https://github.com/OISF/suricata/pull/12070

### CODE REVIEW

File: src/detect-vlan-id.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+#ifndef SURICATA_DETECT_VLAN_ID_H
+#define SURICATA_DETECT_VLAN_ID_H
+
+#define ANY 0
```
Comment (catenacyber) [inline src/detect-vlan-id.h:21]: I do not think we want to export this

File: src/detect-vlan-id.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+typedef struct DetectVlanIdData_ {
+    uint16_t id;   /* id to match */
+    uint8_t layer; /* id layer to match */
+} DetectVlanIdData;
```
Comment (catenacyber) [inline src/detect-vlan-id.h:26]: This structure is only used in detect-vlan-id.c: it can live there

File: src/detect-vlan-id.c [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+#include "detect-engine-prefilter.h"
+#include "detect-engine-uint.h"
+#include "detect-parse.h"
+#include "util-byte.h"
+
+#ifdef UNITTESTS
+static void DetectVlanIdRegisterTests(void);
+#endif
+
+#define PARSE_REGEX "^([0-9]+)(?:,\\s*([0-9]|any))?$"
```
Comment (catenacyber) [inline src/detect-vlan-id.c:31]: I think we do not want anymore parsing with regex, rust would be welcome

File: doc/userguide/rules/vlan-id.rst [hunk 4]
```diff
@@ -0,0 +1,21 @@
+Vlan.id Keyword
+==============
```
Comment (catenacyber) [inline doc/userguide/rules/vlan-id.rst:2]: Miss a `=`

File: doc/userguide/rules/vlan-id.rst [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+ alert ip any any -> any any (msg:"Vlan ID is equal to 300"; vlan.id:300; sid:1;)
+
+::
+
+ alert ip any any -> any any (msg:"Vlan ID is equal to 300"; vlan.id:300,1; sid:1;)
+
+::
+
+ alert ip any any -> any any (msg:"Vlan ID is equal to 200"; vlan.id:200,any; sid:1;)
```
Comment (catenacyber) [inline doc/userguide/rules/vlan-id.rst:21]: Reading the doc, what is the difference between `vlan.id:200,any;` and `vlan.id:200;` ?

File: src/detect-vlan-id.c [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+static int DetectVlanIdMatch(
+        DetectEngineThreadCtx *det_ctx, Packet *p, const Signature *s, const SigMatchCtx *ctx)
+{
+    if (p->vlan_idx == 0) {
+        return 0;
+    }
+
+    const DetectVlanIdData *vdata = (const DetectVlanIdData *)ctx;
+    for (int i = 0; i < p->vlan_idx; i++) {
+        if (p->vlan_id[i] == vdata->id && (vdata->layer == ANY || vdata->layer - 1 == i))
```
Comment (catenacyber) [inline src/detect-vlan-id.c:44]: I would like to use generic integer support here, to support negation for instance.

File: src/detect-vlan-id.c [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+
+static int DetectVlanIdMatch(
+        DetectEngineThreadCtx *det_ctx, Packet *p, const Signature *s, const SigMatchCtx *ctx)
+{
+    if (p->vlan_idx == 0) {
+        return 0;
+    }
+
+    const DetectVlanIdData *vdata = (const DetectVlanIdData *)ctx;
+    for (int i = 0; i < p->vlan_idx; i++) {
```
Comment (catenacyber) [inline src/detect-vlan-id.c:43]: Do not need the loop of vdata->layer != ANY

File: src/detect-vlan-id.c [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+    PASS;
+}
+
+/**
+ * \test DetectVlanIdParseTest03 is a test for setting a valid vlan id value and explicit "any" vlan
+ * layer
+ */
+static int DetectVlanIdParseTest03(void)
+{
+    DetectVlanIdData *vdata = DetectVlanIdParse(NULL, "200,any");
```
Comment (catenacyber) [inline src/detect-vlan-id.c:229]: Should we support `200,last` ? or `200,-1` ?

File: src/detect-vlan-id.h [hunk 9]
```diff
@@ -0,0 +1,30 @@
+/* Copyright (C) 2022 Open Information Security Foundation
```
Comment (catenacyber) [inline src/detect-vlan-id.h:1]: Nit 2024

File: doc/userguide/rules/vlan-id.rst [hunk 10]
```diff
@@ -0,0 +1,21 @@
+Vlan.id Keyword
+==============
+
+Suricata has a ``vlan.id`` keyword that can be used in signatures to identify
+and filter network packets based on Virtual Local Area Network IDs.
+
+Syntax::
+
+ vlan.id: id[,layer];
```
Comment (catenacyber) [inline doc/userguide/rules/vlan-id.rst:9]: What is the default behavior when we do not specify a layer ?

**Review by catenacyber — CHANGES_REQUESTED** [review]
Thanks for the nice work, I think we should get rid of PCRE here

File: doc/userguide/rules/vlan-id.rst (hunk 10 above)
Comment (AkakiAlice) [reply doc/userguide/rules/vlan-id.rst:9]: It behaves just like specifying 'any',  it matches any of the VLAN IDs.

Comment (jufajardini) [reply doc/userguide/rules/vlan-id.rst:9]: Then it's good to have that added to the docs :)

File: src/detect-vlan-id.c [hunk 11]
```diff
@@ -0,0 +1,276 @@
+/* Copyright (C) 2022 Open Information Security Foundation
```
Comment (jufajardini) [inline src/detect-vlan-id.c:1]: nit: copyright year should be 2024

**Pull request closed without merging**